python/setup_project/storage_map_page.py

```diff
@@ -0,0 +1,398 @@
+        roots are cleared so they can be rebuilt with potentially a different
+        configuration.
+        """
+
+        # the wizard instance and its UI
+        wiz = self.wizard()
+        ui = wiz.ui
+
+        # the scroll area widget where the mapping widgets live
+        layout = ui.storage_map_area_widget.layout()
+
+        # reset the internal data
+        self._required_roots = []
+        self._map_widgets = []
+
+        # iterate over all layout items and remove them. if the item is a
+        # storage map widget, delete it.
+        for index in range(0, layout.count()):
+            layout_item = layout.takeAt(0)
+            widget = None
+            if layout_item:
+                widget = layout_item.widget()
+            layout.removeItem(layout_item)
+            if widget and isinstance(widget, StorageMapWidget):
+                widget.deleteLater()
+
+    def set_config_uri(self, uri):
+        """Set the config URI."""
+        self._uri = uri
+
+    def validatePage(self):
+        """The 'next' button was pushed. See if the mappings are valid."""
+
+        # the wizard instance and its UI
+        wiz = self.wizard()
+        ui = wiz.ui
+
+        # clear any errors
+        ui.storage_errors.setText("")
+
+        # get the path key for the current os
+        if sys.platform.startswith("linux"):
+            current_os_key = "linux_path"
+        elif sys.platform == "darwin":
+            current_os_key = "mac_path"
+        elif sys.platform == "win32":
+            current_os_key = "windows_path"
+        else:
+            ui.storage_errors.setText("The current OS is unrecognized.")
+            return False
+
+        # temp lists of widgets that need attention
+        invalid_widgets = []
+        not_on_disk_widgets = []
+
+        # keep track of the first invalid widget so we can ensure it is visible
+        # to the user in the list.
+        first_invalid_widget = None
+
+        # see if each of the mappings is valid
+        for map_widget in self._map_widgets:
+
+            if not map_widget.mapping_is_valid():
+                # something is wrong with this widget's mapping
+                invalid_widgets.append(map_widget)
+                if first_invalid_widget is None:
+                    first_invalid_widget = map_widget
+
+            storage = map_widget.local_storage or {}
+            current_os_path = storage.get(current_os_key)
+
+            if current_os_path and not os.path.exists(current_os_path):
+                # the current os path for this widget doesn't exist on disk
+                not_on_disk_widgets.append(map_widget)
+
+        if invalid_widgets:
+            # tell the user which roots don't have valid mappings
+            root_names = [w.root_name for w in invalid_widgets]
+            ui.storage_errors.setText(
+                "The mappings for these roots are invalid: <b>%s</b>" %
```

Comment:
could we provide more details as log messages?

Reply:
done.